### PR TITLE
fix Last Will (LWT) not set on MQTT Connect #2247

### PR DESCRIPTION
**Pull request by proddy**

### CODE REVIEW

File: lib/framework/MqttSettingsService.cpp [hunk 1]
```diff
@@ -207,6 +207,17 @@ bool MqttSettingsService::configureMqtt() {
...
+        // create last will topic with the base prefixed. It has to be static because the client destroys the reference
+        static char will_topic[FACTORY_MQTT_MAX_TOPIC_LENGTH];
+        if (_state.base.isEmpty()) {
+            snprintf(will_topic, sizeof(will_topic), "status");
+        } else {
+            snprintf(will_topic, sizeof(will_topic), "%s/status", _state.base.c_str());
+        }
+        setWill(will_topic);
+
+
```
Comment (MichaelDvP) [inline lib/framework/MqttSettingsService.cpp:220]: This only effects the unsecure client. 
Better remove `setWill()`function, set the will_topic before line 187 (#ifndef TASMOTA) and use 
`static_cast<espMqttClientSecure *>(_mqttClient)->setWill(will_topic, 1, true, "offline");`
and
`static_cast<espMqttClient *>(_mqttClient)->setWill(will_topic, 1, true, "offline");`
in the secure/unsecure sections (before line 200/210)

**Review by MichaelDvP — CHANGES_REQUESTED** [review]
I think this is the right way and works on the unsecure client, setting for secureclient  is missing.

File: src/mqtt.cpp [hunk 2]
```diff
@@ -524,6 +514,16 @@ void Mqtt::on_connect() {
...
+    // create last will topic with the base prefixed. It has to be static because the client destroys the reference
+    static char will_topic[MQTT_TOPIC_MAX_SIZE];
+    if (!Mqtt::base().empty()) {
+        snprintf(will_topic, MQTT_TOPIC_MAX_SIZE, "%s/status", Mqtt::base().c_str());
+    } else {
+        snprintf(will_topic, MQTT_TOPIC_MAX_SIZE, "status");
+    }
+    // EMSESP::esp8266React.setWill(will_topic); // with qos 1, retain true
+
+
```
[the remark this answers is not in the document]
Comment (proddy) [reply src/mqtt.cpp:526]: oops, it was left in. I was testing whether you had to set the will before or after the connect

File: lib/framework/MqttSettingsService.cpp (hunk 1 above)
Comment (proddy) [reply lib/framework/MqttSettingsService.cpp:220]: setWill does before secure and unsecure, but indeed this approach is cleaner. I'll update

**Pull request merged**